File: js/endless-binary/addition.js
// taidalab Version 1.1.0
// https://github.com/taidalog/taidalab
// Copyright (c) 2022 taidalog
// This software is licensed under the MIT License.
// https://github.com/taidalog/taidalab/blob/main/LICENSE
function checkAnswerAdd(answer, num1, num2, last_answers) {
    
    const numberInput = document.getElementById('numberInput');
    const bin = escapeHtml(numberInput.value);
    console.log(bin);
    
    const errorArea = document.getElementById('errorArea');
    errorArea.innerHTML = '';
    
    const sourceRadix = 2;
    if (bin == '') {
        errorArea.innerHTML = '<span class="warning">' + num1.toString(sourceRadix) + '<small>(' + sourceRadix + ')</small>' + num2.toString(sourceRadix)+ '<small>(' + sourceRadix + ')</small>' + ' の2進法表記を入力してください。</span>';
    } else if (testBinaryString(bin) == false) {
        errorArea.innerHTML = '<span class="warning">"' + bin + '" は2進数ではありません。使えるのは半角の 0 と 1 のみです。</span>';
    } else {
        
        const digit = 8;
        const zeroPaddedBin = bin.padStart(digit, '0');
        const taggedBin = colorLeadingZero(zeroPaddedBin);
        const dec = parseInt(bin, sourceRadix);
        console.log(taggedBin);
        console.log(dec);
        
        const destinationRadix = 10;
        const outputArea = document.getElementById('outputArea');
        const currentHistoryMessage = newHistory((dec == answer), taggedBin, sourceRadix, dec, destinationRadix);
        const historyMessage = concatinateStrings(currentHistoryMessage, outputArea.innerHTML);
        console.log(currentHistoryMessage);
        console.log(historyMessage);
        outputArea.innerHTML = historyMessage;
        
        if (dec == answer) {
            // Making next question.
            let numbers = [];

            console.log(last_answers);
            do {
                numbers = newNumbersAdd();
                console.log(numbers[0]);
                console.log(numbers[1]);
                console.log(last_answers.some((element) => element == numbers[0] || element == numbers[1]));
            } while (last_answers.some((element) => element == numbers[0] || element == numbers[1]));

            console.log(numbers[0]);
            console.log(numbers[1]);
            console.log(numbers[0] + numbers[1]);
            console.log((numbers[0] + numbers[1]).toString(sourceRadix));
            setColumnAddition(numbers[0], numbers[1]);

            const hintArea = document.getElementById('hintArea');
            const nextHint = newHintAdd();
            hintArea.innerHTML = nextHint;
            console.log(nextHint);

            numberInput.value = '';

            const answersToKeep = 20;
            const lastAnswers = [numbers[0], numbers[1]].concat(last_answers).slice(0, answersToKeep);
            document.getElementById('submitButton').onclick = function () { checkAnswerAdd((numbers[0] + numbers[1]), numbers[0], numbers[1], lastAnswers); return false; };
        }
    }
    
    numberInput.focus();
}


function newNumbersAdd () {
    
    const regex = /^1+0+$/;
    let number1 = 0;

    do {
        number1 = getRandomBetween(1, 255);
        console.log('number1: ' + number1);
        console.log('number1.binary: ' + number1.toString(2));
        console.log('number1.length: ' + number1.toString(2).length);
    } while ((number1.toString(2).length == 8) && regex.test(number1.toString(2)));

    let number2 = 0;

    do {
        number2 = getRandomBetween(1, 255 - number1);
        console.debug('number1 == number2: ' + (number1 == number2) + '\t(number1 & number2) == 0: ' + ((number1 & number2) == 0));
    } while ((number1 == number2) || ((number1 & number2) == 0));

    console.log(number2);
    console.log(number1 + number2);
    return [number1, number2];
}


function newHintAdd () {
    const hintFormat01 = '<details><summary>ヒント: </summary>';
    const hintFormat02 = '<p class="history-indented">';
    const hintFormat03 = '10進数の筆算と同じように、右端から上下の数を足していきます。<br><br>';
    const hintFormat04 = '0<sub>(2)</sub> + 0<sub>(2)</sub> = 0<sub>(2)</sub><br>';
    const hintFormat05 = '0<sub>(2)</sub> + 1<sub>(2)</sub> = 1<sub>(2)</sub><br>';
    const hintFormat06 = '1<sub>(2)</sub> + 1<sub>(2)</sub> = 10<sub>(2)</sub><br>';
    const hintFormat07 = '1<sub>(2)</sub> + 1<sub>(2)</sub> + 1<sub>(2)</sub> = 11<sub>(2)</sub><br><br>';
    const hintFormat08 = '10<sub>(2)</sub> や 11<sub>(2)</sub>のように桁が繰り上がった時は、<br>';
    const hintFormat09 = '繰り上がった桁 (=1) をひとつ左の桁に足します。<br>';
    const hintFormat10 = '</p>';
    const hintFormat11 = '</details>';
    const hint = hintFormat01 + hintFormat02 + hintFormat03 + hintFormat04 + hintFormat05 + hintFormat06 + hintFormat07 + hintFormat08 + hintFormat09 + hintFormat10 + hintFormat11;
    return hint;
}


function initAddition () {
    // initialization
    const sourceRadix = 2;
    const destinationRadix = 2;
    const hint = newHintAdd();

    document.getElementById('numberInput').className = 'number-input question-number eight-digit';
    document.getElementById('operator').innerText = '+)';
    document.getElementById('firstRowSrcRadix').innerHTML = '(' + sourceRadix + ')';
    document.getElementById('secondRowSrcRadix').innerHTML = '(' + sourceRadix + ')';
    document.getElementById('binaryRadix').innerHTML = '<sub>(' + destinationRadix + ')</sub>';
    document.getElementById('hintArea').innerHTML = hint;

    const numbers = newNumbersAdd();
    console.log(numbers[0]);
    console.log(numbers[1]);
    console.log(numbers[0] + numbers[1]);
    console.log((numbers[0] + numbers[1]).toString(sourceRadix));
    setColumnAddition(numbers[0], numbers[1]);

    document.getElementById('submitButton').onclick = function () { checkAnswerAdd((numbers[0] + numbers[1]), numbers[0], numbers[1], [numbers[0], numbers[1]]); return false; };
}

File: src/endless-binary/Addition.fs
// taidalab
// https://github.com/taidalog/taidalab
// Copyright (c) 2022-2025 taidalog
// This software is licensed under the MIT License.
// https://github.com/taidalog/taidalab/blob/main/LICENSE
namespace Taidalab

open Browser.Dom
open Browser.Types
open Taidalab.Number
open Taidalab.Text
open Taidalab.EndlessBinary
open Fermata
open Fermata.RadixConversion

module EndlessBinary =
    module Addition =
        let help =
            """
            2進数同士の足し算をエンドレスで練習できます。<br>
            出題範囲は m, n (2 &le; m + n &le; 255) で、繰り上がりもあります。<br>
            ヒント付きなので、考え方も身に付けられます。
            """

        let newHintAdd () =
            let hint =
                """
                <details><summary><h2>ヒント:</h2></summary>
                    <p class="history-indented">
                        10進数の筆算と同じように、右端から上下の数を足していきます。<br><br>
                        0<sub>(2)</sub> + 0<sub>(2)</sub> = 0<sub>(2)</sub><br>
                        0<sub>(2)</sub> + 1<sub>(2)</sub> = 1<sub>(2)</sub><br>
                        1<sub>(2)</sub> + 1<sub>(2)</sub> = 10<sub>(2)</sub><br>
                        1<sub>(2)</sub> + 1<sub>(2)</sub> + 1<sub>(2)</sub> = 11<sub>(2)</sub><br><br>
                        10<sub>(2)</sub> や 11<sub>(2)</sub>のように桁が繰り上がった時は、<br>
                        繰り上がった桁 (=1) をひとつ左の桁に足します。<br>
                    </p>
                </details>"""

            hint

        let newNumbersAdd (digit: int) : (int * int) =
            let max = digit |> pown 2 |> (+) -1

            let number1 =
                newNumber
                    (fun _ -> getRandomBetween 1 max)
                    (if digit < 4 then
                         (fun n ->
                             n
                             |> bin
                             |> function
                                 | Bin v -> v
                             |> String.length = digit)
                     else
                         (fun n ->
                             let pattern = "^1+0+$"
                             let (Bin v) = bin n
                             (String.length v = digit) && (Regex.isMatch pattern v) = false))

            let number2 =
                let max' = if digit < 4 then max else max - number1

                newNumber
                    (fun _ -> getRandomBetween 1 max')
                    (if digit < 4 then
                         (fun _ -> true)
                     else
                         (fun n -> (n <> number1) && ((n &&& number1) <> 0)))

            (number1, number2)

        let question (digit: int) (lastAnswers: int list) : int * int =
            newNumber
                (fun _ -> newNumbersAdd digit)
                (if digit < 4 then
                     (fun _ -> true)
                 else
                     (fun (x, y) -> List.contains x lastAnswers = false && List.contains y lastAnswers = false))

        let questionExpression (x: int, y: int) : string =
            let intToBinExpr (x: int) = let (Bin v) = bin x in v
            $"%s{intToBinExpr x}<sub>(2)</sub> + %s{intToBinExpr y}<sub>(2)</sub>"

        let history (correct: bool) (input: string) : string =
            match input |> Bin.validate with
            | Error _ -> ""
            | Ok v ->
                let (Dec d) = Bin.toDec v
                let left = input |> Fermata.String.padLeft 8 ' ' |> escapeSpace
                let right = d |> string |> Fermata.String.padLeft 4 ' ' |> escapeSpace
                newHistory correct left 2 right 10

        let rec exec
            (questionf: int list -> int * int)
            (questionExpressionf: int * int -> string)
            (hintf: unit -> string)
            (collumnf: int -> int -> unit)
            (answerf: int -> int -> int)
            (answersToKeep: int)
            (lastAnswers: int list)
            (num1: int)
            (num2: int)
            (answer: int)
            : unit =
            // Getting the user input.
            let numberInput = document.getElementById "numberInput" :?> HTMLInputElement
            let input: string = numberInput.value |> escapeHtml
            let input': Result<Bin, exn> = input |> Bin.validate

            numberInput.focus ()

            match input' with
            | Error e ->
                // Making an error message.
                (document.getElementById "errorArea").innerHTML <-
                    newErrorMessageBin (questionExpressionf (num1, num2)) input e
            | Ok v ->
                (document.getElementById "errorArea").innerHTML <- ""

                let (Dec d) = Bin.toDec v

                // Making a new history and updating the history with the new one.
                let outputArea = document.getElementById "outputArea"

                let historyMessage =
                    history (d = answer) input
                    |> (fun x -> concatinateStrings "<br>" [ x; outputArea.innerHTML ])

                outputArea.innerHTML <- historyMessage

                if d = answer then
                    // Making the next question.
                    let (number1, number2) = questionf lastAnswers

                    collumnf number1 number2

                    (document.getElementById "hintArea").innerHTML <- hintf ()

                    numberInput.value <- ""

                    // Updating `lastAnswers`.
                    // These numbers will not be used for the next question.
                    let lastAnswers' =
                        ([ number1; number2 ] @ lastAnswers) |> List.truncate answersToKeep

                    // Setting the next answer to the check button.
                    let f =
                        fun (e: Event) ->
                            e.preventDefault ()

                            exec
                                questionf
                                questionExpressionf
                                hintf
                                collumnf
                                answerf
                                answersToKeep
                                lastAnswers'
                                number1
                                number2
                                (answerf number1 number2)

                    (document.getElementById "submitButton").onclick <- f
                    (document.getElementById "inputArea").onsubmit <- f


        let init'
            (questionf: int list -> int * int)
            (questionExpressionf: int * int -> string)
            (hintf: unit -> string)
            (sourceRadix: int)
            (destinationRadix: int)
            (answersToKeep: int)
            (keyboardshortcut: KeyboardEvent -> unit)
            executor
            : unit =
            // Initialization.
            (document.getElementById "numberInput").className <- "question-number"
            (document.getElementById "operator").innerText <- "+)"
            (document.getElementById "firstRowSrcRadix").innerText <- sprintf "(%d)" sourceRadix
            (document.getElementById "secondRowSrcRadix").innerText <- sprintf "(%d)" sourceRadix
            (document.getElementById "binaryRadix").innerHTML <- sprintf "<sub>(%d)</sub>" destinationRadix
            (document.getElementById "hintArea").innerHTML <- hintf ()

            let (number1, number2) = questionf []
            setColumnAddition number1 number2

            let f =
                fun (e: Event) ->
                    e.preventDefault ()

                    executor
                        questionf
                        questionExpressionf
                        hintf
                        setColumnAddition
                        (+)
                        answersToKeep
                        [ number1; number2 ]
                        number1
                        number2
                        (number1 + number2)

            (document.getElementById "submitButton").onclick <- f
            (document.getElementById "inputArea").onsubmit <- f

            (document.getElementById "helpButton").onclick <-
                (fun _ ->
                    [ "helpWindow"; "helpBarrier" ]
                    |> List.iter (fun x -> (document.getElementById x).classList.toggle "active" |> ignore))

            (document.getElementById "helpBarrier").onclick <-
                (fun _ ->
                    [ "helpWindow"; "helpBarrier" ]
                    |> List.iter (fun x -> (document.getElementById x).classList.remove "active" |> ignore))

            (document.getElementById "helpClose").onclick <-
                (fun _ ->
                    [ "helpWindow"; "helpBarrier" ]
                    |> List.iter (fun x -> (document.getElementById x).classList.remove "active" |> ignore))

            document.onkeydown <- (fun (e: KeyboardEvent) -> keyboardshortcut e)

        let init () =
            document.title <- "加算 - taidalab"

            let header = document.querySelector "header"
            header.innerHTML <- Content.Common.header
            header.className <- "addition"

            (document.getElementById "hamburgerButton").onclick <-
                (fun _ ->
                    (document.querySelector "nav").classList.toggle "flagged" |> ignore
                    (document.getElementById "barrier").classList.toggle "flagged" |> ignore
                    (document.querySelector "main").classList.toggle "flagged" |> ignore)

            (document.getElementById "barrier").onclick <-
                (fun _ ->
                    (document.querySelector "nav").classList.remove "flagged" |> ignore
                    (document.getElementById "barrier").classList.remove "flagged" |> ignore
                    (document.querySelector "main").classList.remove "flagged" |> ignore)

            (document.querySelector "#headerTitle").innerHTML <-
                """<span>加算 - </span><span translate="no">taidalab</span>"""

            (document.querySelector "main").innerHTML <- EndlessBinary.Course.main help "help-color addition"
            (document.querySelector "#submitButton").className <- "addition"
            (document.querySelector "#questionArea").innerHTML <- Content.Common.columnAdditionFormat

            init' (question 8) questionExpression newHintAdd 2 2 10 EndlessBinary.keyboardshortcut exec

        let init4 () =
            init' (question 4) questionExpression newHintAdd 2 2 5 EndlessBinary.keyboardshortcut exec
